packages/transactions/src/builders.ts

```diff
@@ -139,11 +139,20 @@ export async function estimateTransfer(
   return feeRate.mul(txBytes);
 }
 
-export type TxBroadcastResultOk = string;
+export type TxBroadcastResultOk = {
+  txid: string;
+};
 export type TxBroadcastResultRejected = {
   error: string;
   reason: TxRejectedReason;
-  reason_data: any;
+  reason_data: {
+    message?: string;
+    expected?: string | number;
+    actual?: string | number;
+    is_origin?: boolean;
+    principal?: string;
+    contract_identifier?: string;
+  };
```

Comment:
I think fully typed errors would be a type union like, something like:
```ts
interface SerializationRejection {
  reason: 'Serialization';
  reason_data: {
    message: string;
  }
}

interface BadNonceRejection {
  reason: 'BadNonce';
  reason_data: {
    expected: number;
    actual: number;
    is_origin: boolean;
    principal: boolean;
  }
}

type TxBroadcastResultRejected = SerializationRejection | BadNonceRejection | (all the other ones...);
```

Reply:
@zone117x Added these type changes. Give it a look when you get a chance.

Reply:
Wow, now that's typed! One lasts question around how to easily check if a `TxBroadcastResult` result is success vs an error. For example, looking at the final type union, it doesn't look like it's possible to do simple check without typecasting, something like this doesn't seem like it would work:
```ts
const result: TxBroadcastResult;
if (result.error) {
  throw new Error(...);
}
```

Reply:
@zone117x Exactly to check the error the result needs to be type casted to TxBroadcastResultRejected
Example: 
```
  const reply: TxBroadcastResult = await broadcastTransaction(
      transaction, network
  );
  if ((reply as TxBroadcastResultRejected).error) {
   // throw new Error(...);
    console.log('Reason: ', (reply as TxBroadcastResultRejected).reason);
  } else {
    console.log('Txid: ', reply.txid);
  }

```
I spent some time find out some easy way to achieve it in a way as you mentioned above. But eventually we have to typecast it to exact type to check the error or reason property. Other workaround would be be generalize the TxBroadcastResult keeping all the properties of TxBroadcastResultRejected but that might not be a good way. In case you have any suggestion to improve this solution, let me know so that i can incorporate it to make it better.

Reply:
In the past, I've used a `error: null` property on the success object (assuming all the failure types have a `error: <truthy type>` property.

Reply:
@zone117x Let's talk about what's currently have in master branch. Here is the snapshot of types:
```
type TxBroadcastResultOk = string;
type TxBroadcastResultRejected = {
  error: string;
  reason: TxRejectedReason;
  reason_data: any;
  txid: string;
}
type TxBroadcastResult = TxBroadcastResultOk | TxBroadcastResultRejected;
```
You can see here that TxBroadcastResult will be string in case of success. So i am wondering now how you used `error: null` on string type in past.

Reply:
Not in this codebase -- I've used the practice of typing `result` objects with a `interface success { error?: undefined }` in the past so that the type union can easily be used.

Reply:
@zone117x Not sure about the other codebase that you are talking about. If you can mention link to your other codebase or example types to achieve that scenario that would be great. 

@kyranjamie Adding you in this discussion. Let me know if you have any ideas how we can achieve as @zone117x mentioned above.

Reply:
Sure, here's an example I just tested and seems to work as expected:
```ts
export type TxBroadcastResultOk = {
  txid: string;
  error?: undefined;
};

export const example: TxBroadcastResult = null as any;

if (example.error) {
  switch (example.reason) {
    // Example of special handling for a specific error type
    case TxRejectedReason.BadNonce:
      alert(
        `Nonce expected: ${example.reason_data.expected}, actual: ${example.reason_data.actual}`
      );
      throw new Error(example.error);
    default:
      throw new Error(example.error);
  }
} else {
  console.log('success: ' + example.txid);
}
```